File: server/utils/di.go
package utils

import (
	"fmt"
)

var ErrKeyAlreadyExists = fmt.Errorf("map: key already exists")

type CreateFunc func(args ...interface{}) any

type di_config struct {
	is_only     bool
	instance    interface{}
	create_func CreateFunc
}

type DI interface {
	Set(key string, create_func CreateFunc) (e error)
	SetShare(key string, create_func CreateFunc) (e error)
	MustGet(key string, args ...interface{}) any
}

//-----------------------------------------------

type di_store map[string]di_config

func (s di_store) Set(key string, create_func CreateFunc) (e error) {
	_, ok := s[key]
	if ok {
		return ErrKeyAlreadyExists
	}
	s[key] = di_config{
		is_only:     false,
		create_func: create_func,
	}
	return
}

func (s di_store) SetShare(key string, create_func CreateFunc) (e error) {
	_, ok := s[key]
	if ok {
		return ErrKeyAlreadyExists
	}
	s[key] = di_config{
		is_only:     true,
		create_func: create_func,
	}
	return
}

func (s di_store) MustGet(key string, args ...interface{}) any {
	config, ok := s[key]
	if !ok {
		panic(fmt.Errorf("service(%s) unset", key))
	}

	if config.is_only {
		if config.instance == nil {
			config.instance = config.create_func(args...)
		}
		return config.instance
	}

	return config.create_func(args...)
}

//-----------------------------------------------

var di = &di_store{}

func GetDI() DI {
	return di
}

File: client/ccs/Demo/assets/scripts/components/DI.ts
export class DIKey {
    static WSConn: string = "websocket.conn";
    static WSObserver: string = "websocket.observer";
    static LOGGER: string = "logger";
}
class di_config {
    is_only: boolean;
    instance: any;
    create_func: CreateFunc;
}
export type CreateFunc = (...args: any) => any;
export class DI {
    private static store: { [key: string]: di_config } = {};
    public static set(key: string, create_func: CreateFunc) {
        if (this.store[key] != undefined) {
            throw new Error(`Key: ${key} already exists.`);
        }
        let config = new di_config();
        config.is_only = false;
        config.create_func = create_func;
        this.store[key] = config;
    }
    public static set_share(key: string, create_func: CreateFunc) {
        if (this.store[key] != undefined) {
            throw new Error(`Key: ${key} already exists.`);
        }
        let config = new di_config();
        config.is_only = true;
        config.create_func = create_func;
        this.store[key] = config;
    }
    public static must_get<T>(key: string, ...args): T {
        if (this.store[key] == undefined) {
            throw new Error(`Key: ${key} has an empty value.`);
        }
        let config = this.store[key];
        if (config.is_only) {
            if (config.instance == undefined) {
                config.instance = config.create_func(...args) ;
            }
            return config.instance ;
        }
        return <T>config.create_func(...args) ;
    }
}
